backend/services/implementations/shiftService.ts

```diff
@@ -0,0 +1,372 @@
+    if (startTime.format(dateFormat) !== endTime.format(dateFormat)) {
+      return [false, "startTime and endTime must be in the same day"];
+    }
+    return [true, ""];
+  }
+
+  validateTimeBlocks(timeBlocks: TimeBlockDTO[]): [boolean, string] {
+    if (timeBlocks.length === 0) return [false, "No time blocks provided"];
+    // Check that start time is before end time
+    if (
+      timeBlocks.some(
+        (tb) =>
+          !tb.date ||
+          !tb.startTime ||
+          !tb.endTime ||
+          !moment(
+            `${tb.date} ${tb.startTime}`,
+            dateTimeFormat,
+            true,
+          ).isValid() ||
+          !moment(
+            `${tb.date} ${tb.endTime}`,
+            dateTimeFormat,
+            STRICT_MODE,
+          ).isValid() ||
+          moment(tb.startTime, timeFormat) >= moment(tb.endTime, timeFormat),
+      )
+    )
+      return [false, "Invalid time blocks"];
+
+    const [earliestDate, latestDate] = timeBlocks.reduce(
+      (earliestLatestSoFar, currentTimeblock) => {
+        const earliestLatestTuple = earliestLatestSoFar;
+        const currentDate = moment(
+          currentTimeblock.date,
+          dateFormat,
+          STRICT_MODE,
+        );
+        earliestLatestTuple[0] = currentDate.isBefore(earliestLatestSoFar[0])
+          ? currentDate
+          : earliestLatestSoFar[0];
+        earliestLatestTuple[1] = currentDate.isAfter(earliestLatestSoFar[1])
+          ? currentDate
+          : earliestLatestSoFar[1];
+        return earliestLatestTuple;
+      },
+      [
+        moment(timeBlocks[0].date, dateFormat, STRICT_MODE),
+        moment(timeBlocks[0].date, dateFormat, STRICT_MODE),
+      ],
+    );
+
+    if (moment.duration(latestDate.diff(earliestDate)).asDays() >= 7)
+      return [false, "Time blocks must be within a week"];
+
+    return [true, ""];
+  }
+
+  buildTimeBlocks(shifts: ShiftBulkRequestDTO): TimeBlock[] {
+    const endDate: Moment = moment(shifts.endDate, dateFormat, STRICT_MODE).add(
+      1,
+      "day",
+    );
+
+    // Get moment's duration function args using recurrence
+    const duration: DurationArgs | null = this.getDuration(
+      shifts.recurrenceInterval,
+    );
+
+    const shiftTimes: TimeBlock[] = shifts.times.flatMap((time) => {
+      const startTime = moment(
+        `${time.date} ${time.startTime}`,
+        dateTimeFormat,
+        true,
+      );
+      const endTime = moment(
+        `${time.date} ${time.endTime}`,
+        dateTimeFormat,
+        true,
+      );
+      const recurringShifts = [];
+      let end = endTime;
+      if (duration) {
+        for (
+          let start = startTime.clone();
+          start < endDate;
+          start.add(duration.value, duration.unit)
+        ) {
+          recurringShifts.push({
+            startTime: start.toDate(),
+            endTime: end.toDate(),
+          });
+          end = end.add(duration.value, duration.unit);
+        }
```

Comment:
Can declare and increment both start and end in loop initialization. Side note: is the `startTime.clone()` necessary?

```suggestion
        for (
          let start = startTime.clone(), end = endTime;
          start < endDate;
          start.add(duration.value, duration.unit), end.add(duration.value, duration.unit)
        ) {
          recurringShifts.push({
            startTime: start.toDate(),
            endTime: end.toDate(),
          });
        }
```